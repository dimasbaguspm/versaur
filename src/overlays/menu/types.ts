import type { HTMLAttributes, MouseEvent, ReactNode, RefObject } from 'react'

export type MenuSize = 'sm' | 'md'
export type MenuPlacement =
  | 'bottom-start'
  | 'bottom-end'
  | 'top-start'
  | 'top-end'
  | 'auto'

/**
 * Props for Menu
 */
export interface MenuProps {
  /** Whether the menu is open (controlled externally) */
  isOpen: boolean
  /** Callback when clicking outside menu */
  onOutsideClick: () => void
  /** Menu size variant */
  size?: MenuSize
  /** Menu content (MenuContent/MenuItem) */
  content: ReactNode
  /** Trigger element */
  children: ReactNode
  /** Preferred placement of the menu relative to trigger */
  placement?: MenuPlacement
  /** Container element to respect boundaries (defaults to viewport) */
  container?: HTMLElement | RefObject<HTMLElement | null> | null
  /** Whether to keep the menu after list item clicked */
  preserve?: boolean
}

/**
 * Props for MenuContent
 */
export interface MenuContentProps extends HTMLAttributes<HTMLUListElement> {
  /** Menu items */
  children: ReactNode
}

/**
 * Props for MenuItem
 */
export interface MenuItemProps
  extends Omit<HTMLAttributes<HTMLLIElement>, 'onClick'> {
  /** Item content */
  children: ReactNode
  /** Whether the item is disabled */
  disabled?: boolean
  onClick?: (ev: MouseEvent<HTMLButtonElement>) => void
}
