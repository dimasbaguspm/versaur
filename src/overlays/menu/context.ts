import { createContext, useContext } from 'react'
import type { MenuSize } from './types'

export interface MenuContextValue {
  /** Whether the menu is open (controlled externally) */
  open: boolean
  /** Ref to the trigger button */
  triggerRef: React.RefObject<HTMLSpanElement | null>
  /** Ref to the menu content */
  contentRef: React.RefObject<HTMLDivElement | null>
  /** Menu size variant */
  size: MenuSize
}

export const MenuContext = createContext<MenuContextValue | null>(null)

export const useMenuContext = (): MenuContextValue => {
  const context = useContext(MenuContext)
  if (!context) {
    throw new Error('useMenuContext must be used within a Menu component')
  }
  return context
}
