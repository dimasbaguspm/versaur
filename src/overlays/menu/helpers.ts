/**
 * Utility for menu positioning and variants
 * Supports Spenicle palette and Versaur design system
 * Follows the pattern used in badge/helpers.ts for consistency
 */
import { cva } from '@/utils/variants'

export const menuVariants = cva(
  'absolute z-30 min-w-40 bg-background rounded-lg border border-border transition-all duration-200 ease-out will-change-transform',
  {
    variants: {
      size: {
        sm: 'py-1.5 px-1',
        md: 'py-2 px-1',
      },
      open: {
        true: 'opacity-100 translate-y-1',
        false: 'opacity-0 pointer-events-none translate-y-3',
      },
    },
    defaultVariants: {
      size: 'md',
      open: false,
    },
  }
)

export const menuItemVariants = cva(
  'w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light '
)
