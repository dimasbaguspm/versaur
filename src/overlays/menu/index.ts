export { Menu } from './menu'
export type { MenuProps, MenuSize } from './types'
