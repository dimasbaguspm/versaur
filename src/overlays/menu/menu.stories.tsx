/**
 * Menu stories for Storybook
 * Group: Menu
 */

import { useState } from 'react'
import { Menu } from './menu'
import { MenuContent, MenuItem } from './menu.atoms'
import { ButtonIcon } from '@/primitive/button-icon'
import { Ellipsis, EllipsisVertical } from 'lucide-react'

export default {
  title: 'Overlays/Menu',
  component: Menu,
  parameters: {
    docs: {
      description: {
        component: 'Material Design inspired menu, compound/context pattern.',
      },
    },
  },
}

export const Basic = () => {
  const [isOpen, setIsOpen] = useState(false)
  return (
    <Menu
      isOpen={isOpen}
      onOutsideClick={() => setIsOpen(false)}
      size='md'
      content={
        <MenuContent>
          <MenuItem>Profile</MenuItem>
          <MenuItem>Settings</MenuItem>
          <MenuItem disabled>Logout</MenuItem>
        </MenuContent>
      }
    >
      <ButtonIcon
        as={EllipsisVertical}
        aria-label='Open menu'
        onClick={() => setIsOpen(o => !o)}
      />
    </Menu>
  )
}

export const Sizes = () => {
  const [openSm, setOpenSm] = useState(false)
  const [openMd, setOpenMd] = useState(false)

  return (
    <div className='flex flex-col gap-4'>
      <Menu
        isOpen={openSm}
        onOutsideClick={() => setOpenSm(false)}
        size='sm'
        content={
          <MenuContent>
            <MenuItem>Item 1</MenuItem>
            <MenuItem>Item 2</MenuItem>
          </MenuContent>
        }
      >
        <ButtonIcon
          as={EllipsisVertical}
          aria-label='Open small menu'
          onClick={() => setOpenSm(o => !o)}
        />
      </Menu>
      <Menu
        isOpen={openMd}
        onOutsideClick={() => setOpenMd(false)}
        size='md'
        content={
          <MenuContent>
            <MenuItem>Item 1</MenuItem>
            <MenuItem>Item 2</MenuItem>
          </MenuContent>
        }
      >
        <ButtonIcon
          as={Ellipsis}
          aria-label='Open medium menu'
          onClick={() => setOpenMd(o => !o)}
        />
      </Menu>
    </div>
  )
}
