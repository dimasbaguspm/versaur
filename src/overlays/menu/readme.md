# Menu Component

Accessible, animated menu for Versaur UI. Follows Material Design menu principles. Now uses a simple
API similar to Tooltip.

## Usage

````markdown
# Menu Component

Accessible, animated menu for Versaur UI. Follows Material Design menu principles with smart
positioning to prevent overlaps.

## Usage

```tsx
import { useState } from 'react'
import { Menu, MenuContent, MenuItem } from '@/components/menu'

export default function Example() {
  const [isOpen, setIsOpen] = useState(false)
  return (
    <Menu
      isOpen={isOpen}
      onOutsideClick={() => setIsOpen(false)}
      size='md'
      placement='auto'
      content={
        <MenuContent>
          <MenuItem>Item 1</MenuItem>
          <MenuItem>Item 2</MenuItem>
        </MenuContent>
      }
    >
      <button onClick={() => setIsOpen(o => !o)}>Open Menu</button>
    </Menu>
  )
}
```

## Positioning

The menu supports intelligent positioning to prevent overlaps:

```tsx
// Auto placement - menu chooses best position
<Menu placement='auto' />

// Specific placements
<Menu placement='bottom-start' />
<Menu placement='bottom-end' />
<Menu placement='top-start' />
<Menu placement='top-end' />

// Respect container boundaries
<Menu container={containerRef.current} />
```

## Features

- Smart positioning that respects container boundaries
- Auto placement that chooses optimal position
- Accessible, ARIA roles
- Focus management
- Animated open/close (scale/fade)
- Strictly typed props
- Tailwind styling
- Easy to extend for submenus
````

## Features

- Accessible, ARIA roles
- Focus management
- Animated open/close (slide/fade)
- Strictly typed props
- Tailwind styling
- Easy to extend for submenus
