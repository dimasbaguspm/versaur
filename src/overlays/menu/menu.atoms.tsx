/**
 * Menu atoms for Versaur
 */
import { forwardRef } from 'react'
import type { MenuContentProps, MenuItemProps } from './types'
import { cn } from '@/utils/cn'
import { menuItemVariants } from './helpers'

/**
 * MenuContent: Wraps menu items
 */
export const MenuContent = forwardRef<HTMLUListElement, MenuContentProps>(
  ({ children }, ref) => {
    return (
      <ul ref={ref} className='flex flex-col gap-1'>
        {children}
      </ul>
    )
  }
)

/**
 * MenuItem: Single menu item
 */
export const MenuItem = forwardRef<HTMLLIElement, MenuItemProps>(
  ({ className, ...props }, ref) => {
    return (
      <li ref={ref}>
        <button
          className={cn(menuItemVariants(), className)}
          tabIndex={-1}
          role='menuitem'
          {...props}
        />
      </li>
    )
  }
)
