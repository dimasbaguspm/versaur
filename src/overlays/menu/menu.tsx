import React, { useRef, useId, cloneElement } from 'react'

import { cn } from '@/utils/cn'
import { menuVariants } from './helpers'
import type { MenuProps } from './types'
import {
  useMenuOutsideClick,
  useMenuPosition,
  useMenuEscapeClose,
} from './use-menu'
import { MenuContent, MenuItem } from './menu.atoms'

const MenuRoot: React.FC<MenuProps> = ({
  isOpen,
  onOutsideClick,
  size = 'md',
  content,
  children,
  placement = 'bottom-start',
  container,
}) => {
  const triggerRef = useRef<HTMLButtonElement | null>(null)
  const contentRef = useRef<HTMLDivElement | null>(null)
  const menuId = useId()

  useMenuOutsideClick(isOpen, contentRef, triggerRef, onOutsideClick)
  useMenuEscapeClose(isOpen, onOutsideClick)

  const position = useMenuPosition(
    isOpen,
    triggerRef,
    contentRef,
    placement,
    container
  )

  const positionStyles: React.CSSProperties = {
    ...position,
    overflowY: position.maxHeight ? 'auto' : undefined,
    overflowX: position.maxWidth ? 'auto' : undefined,
  }

  return (
    <div className='relative w-fit'>
      {cloneElement(children as React.ReactElement, {
        // @ts-expect-error: ref is valid for button or forwardRef components
        ref: triggerRef,
        'aria-haspopup': 'menu',
        'aria-expanded': isOpen,
        'aria-controls': menuId,
      })}
      {isOpen && (
        <div
          id={menuId}
          ref={contentRef}
          className={cn(menuVariants({ size, open: isOpen }))}
          style={positionStyles}
          role='menu'
          aria-hidden={!isOpen}
        >
          {content}
        </div>
      )}
    </div>
  )
}

export const Menu = Object.assign(MenuRoot, {
  Content: MenuContent,
  Item: MenuItem,
})
