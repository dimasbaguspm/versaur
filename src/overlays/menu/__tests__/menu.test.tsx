import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import * as stories from '../menu.stories'
import { composeStories } from '@storybook/react'

const { Basic, Placements } = composeStories(stories)

describe('Menu', () => {
  it('renders basic menu correctly', () => {
    const { asFragment } = render(<Basic />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders menu with different placements', () => {
    const { asFragment } = render(<Placements />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('disables menu item', () => {
    render(<Basic />)
    const trigger = screen.getByLabelText('Open menu')
    fireEvent.click(trigger)
    expect(screen.getByText('Logout')).toBeDisabled()
  })

  it('opens and closes menu on trigger click', async () => {
    render(<Basic />)
    const trigger = screen.getByLabelText('Open menu')

    // Menu content exists in DOM but is hidden with Popover API
    const menuPopover = document.querySelector('[popover]')
    expect(menuPopover).toBeInTheDocument()

    // Open menu
    fireEvent.click(trigger)
    await waitFor(() => {
      expect(screen.getByText('Profile')).toBeVisible()
    })
  })

  it('handles menu item clicks', async () => {
    const { getByLabelText, getByText } = render(<Basic />)

    // Open menu
    const trigger = getByLabelText('Open menu')
    fireEvent.click(trigger)

    await waitFor(() => {
      expect(getByText('Profile')).toBeInTheDocument()
    })
  })
})
