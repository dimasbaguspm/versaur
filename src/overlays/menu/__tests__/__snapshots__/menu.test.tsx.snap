// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Menu > renders basic menu correctly 1`] = `
<DocumentFragment>
  <div
    class="relative w-fit"
  >
    <button
      aria-controls="«r0»"
      aria-disabled="false"
      aria-expanded="false"
      aria-haspopup="menu"
      aria-label="Open menu"
      class="inline-flex items-center justify-center font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 w-10 rounded-md"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-ellipsis-vertical inline-flex items-center justify-center w-5 h-5 text-neutral"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="12"
          cy="12"
          r="1"
        />
        <circle
          cx="12"
          cy="5"
          r="1"
        />
        <circle
          cx="12"
          cy="19"
          r="1"
        />
      </svg>
    </button>
    <div
      aria-hidden="true"
      class="absolute z-30 min-w-40 bg-background rounded-lg border border-border transition-all duration-200 ease-out will-change-transform py-2 px-1 opacity-0 pointer-events-none translate-y-3"
      id="«r0»"
      role="menu"
    >
      <ul
        class="flex flex-col gap-1"
      >
        <li>
          <button
            class="w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light"
            role="menuitem"
            tabindex="-1"
          >
            Profile
          </button>
        </li>
        <li>
          <button
            class="w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light"
            role="menuitem"
            tabindex="-1"
          >
            Settings
          </button>
        </li>
        <li>
          <button
            class="w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light"
            disabled=""
            role="menuitem"
            tabindex="-1"
          >
            Logout
          </button>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Menu > renders menu with different sizes 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4"
  >
    <div
      class="relative w-fit"
    >
      <button
        aria-controls="«r1»"
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="Open small menu"
        class="inline-flex items-center justify-center font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 w-10 rounded-md"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-ellipsis-vertical inline-flex items-center justify-center w-5 h-5 text-neutral"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="1"
          />
          <circle
            cx="12"
            cy="5"
            r="1"
          />
          <circle
            cx="12"
            cy="19"
            r="1"
          />
        </svg>
      </button>
      <div
        aria-hidden="true"
        class="absolute z-30 min-w-40 bg-background rounded-lg border border-border transition-all duration-200 ease-out will-change-transform py-1.5 px-1 opacity-0 pointer-events-none translate-y-3"
        id="«r1»"
        role="menu"
      >
        <ul
          class="flex flex-col gap-1"
        >
          <li>
            <button
              class="w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light"
              role="menuitem"
              tabindex="-1"
            >
              Item 1
            </button>
          </li>
          <li>
            <button
              class="w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light"
              role="menuitem"
              tabindex="-1"
            >
              Item 2
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="relative w-fit"
    >
      <button
        aria-controls="«r2»"
        aria-disabled="false"
        aria-expanded="false"
        aria-haspopup="menu"
        aria-label="Open medium menu"
        class="inline-flex items-center justify-center font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-soft focus-visible:ring-offset-white shadow-sm hover:shadow-md h-10 w-10 rounded-md"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-ellipsis inline-flex items-center justify-center w-5 h-5 text-neutral"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="1"
          />
          <circle
            cx="19"
            cy="12"
            r="1"
          />
          <circle
            cx="5"
            cy="12"
            r="1"
          />
        </svg>
      </button>
      <div
        aria-hidden="true"
        class="absolute z-30 min-w-40 bg-background rounded-lg border border-border transition-all duration-200 ease-out will-change-transform py-2 px-1 opacity-0 pointer-events-none translate-y-3"
        id="«r2»"
        role="menu"
      >
        <ul
          class="flex flex-col gap-1"
        >
          <li>
            <button
              class="w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light"
              role="menuitem"
              tabindex="-1"
            >
              Item 1
            </button>
          </li>
          <li>
            <button
              class="w-full text-sm text-left px-2 py-1 rounded-md cursor-pointer duration-150 select-none hover:bg-background-bold outline-none border border-transparent focus-visible:border-primary-light disabled:opacity-50 disabled:cursor-not-allowed disabled:bg-background active:bg-background-light"
              role="menuitem"
              tabindex="-1"
            >
              Item 2
            </button>
          </li>
        </ul>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
